dex/src/state.rs

```diff
@@ -2686,8 +2680,168 @@ impl State {
+
+        let abort = match instruction.side {
+            Side::Bid if coin_credit < instruction.min_coin_qty => true,
+            Side::Ask if native_pc_credit < instruction.min_native_pc_qty => true,
+            _ => false,
+        };
+
+        if abort {
+            solana_program::msg!("Min amount requested not met! Aborting");
+            return Err(DexErrorCode::MinAmountNotMet.into());
+        };
+
+        // Amount that user deposits into the program
+        let deposit_amount;
+        // Amount that user receives after the exchange 
+        let withdraw_amount;
+
+        // Token accounts for transfers
+        let deposit_vault;
+        let withdraw_vault;
+        let deposit_source;
+        let withdraw_destination;
+
+        match instruction.side {
+            Side::Bid => {
+                deposit_amount = native_pc_debit;
+                withdraw_amount = coin_credit.checked_mul(coin_lot_size).unwrap();
+                deposit_vault = pc_vault.token_account();
+                deposit_source = pc_wallet.token_account();
+                withdraw_destination = coin_wallet.token_account();
+                withdraw_vault = coin_vault.token_account();
+
+                market_state.pc_deposits_total += deposit_amount;
+                check_assert!(market_state.coin_deposits_total >= withdraw_amount)?;
+                market_state.coin_deposits_total -= withdraw_amount;
+            }
+            Side::Ask => {
+                deposit_amount = coin_debit.checked_mul(coin_lot_size).unwrap();
+                withdraw_amount = native_pc_credit;
+                deposit_vault = coin_vault.token_account();
+                deposit_source = coin_wallet.token_account();
+                withdraw_destination = pc_wallet.token_account();
+                withdraw_vault = pc_vault.token_account();
+
+                market_state.coin_deposits_total += deposit_amount;
+                check_assert!(market_state.pc_deposits_total >= withdraw_amount)?;
+                market_state.pc_deposits_total -= withdraw_amount;
+            }
+        };
+
+        if deposit_amount != 0 {
+            let balance_before = deposit_vault.balance()?;
+            let deposit_instruction = spl_token::instruction::transfer(
+                &spl_token::ID,
+                deposit_source.inner().key,
+                deposit_vault.inner().key,
+                signer.inner().key,
+                &[],
+                deposit_amount,
+            )
+            .unwrap();
+            invoke_spl_token(
+                &deposit_instruction,
+                &[
+                    deposit_source.inner().clone(),
+                    deposit_vault.inner().clone(),
+                    signer.inner().clone(),
+                    spl_token_program.inner().clone(),
+                ],
+                &[],
+            )
+            .map_err(|err| match err {
+                ProgramError::Custom(i) => match TokenError::from_u32(i) {
```

Comment:
does this actually work? I thought invoke failed unrecoverably on cpi instruction error